File: app/Models/CallRecord.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class CallRecord extends Model
{
    use HasFactory;

    public $timestamps = false;

    protected $fillable = [
        'uuid', 'sip_account_id', 'user_id', 'reseller_id', 'call_flow',
        'caller', 'callee', 'caller_id', 'src_ip', 'dst_ip',
        'incoming_trunk_id', 'outgoing_trunk_id', 'did_id', 'destination_sip_account_id',
        'destination', 'matched_prefix', 'rate_per_minute', 'connection_fee', 'rate_group_id',
        'call_start', 'call_end', 'duration', 'billsec', 'billable_duration',
        'total_cost', 'reseller_cost',
        'disposition', 'hangup_cause', 'status',
        'ast_channel', 'ast_dstchannel', 'ast_context', 'rated_at',
    ];

    protected function casts(): array
    {
        return [
            'call_start' => 'datetime',
            'call_end' => 'datetime',
            'rated_at' => 'datetime',
            'total_cost' => 'decimal:4',
            'reseller_cost' => 'decimal:4',
            'rate_per_minute' => 'decimal:6',
        ];
    }

    /**
     * Datetimes are stored in UTC (the Python engine writes UTC) but displayed
     * in the configured local timezone. The Carbon instance stays TZ-aware, so
     * Carbon math (diffForHumans, comparisons) still works against now().
     */
    protected function asDateTime($value)
    {
        $carbon = parent::asDateTime($value);

        return $carbon
            ? $carbon->setTimezone(config('app.display_timezone', config('app.timezone', 'UTC')))
            : $carbon;
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function reseller(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reseller_id');
    }

    public function sipAccount(): BelongsTo
    {
        return $this->belongsTo(SipAccount::class);
    }

    public function incomingTrunk(): BelongsTo
    {
        return $this->belongsTo(Trunk::class, 'incoming_trunk_id');
    }

    public function outgoingTrunk(): BelongsTo
    {
        return $this->belongsTo(Trunk::class, 'outgoing_trunk_id');
    }

    public function did(): BelongsTo
    {
        return $this->belongsTo(Did::class);
    }

    public function destinationSipAccount(): BelongsTo
    {
        return $this->belongsTo(SipAccount::class, 'destination_sip_account_id');
    }

    public function scopeAnswered($query)
    {
        return $query->where('disposition', 'ANSWERED');
    }

    public function scopeRated($query)
    {
        return $query->where('status', 'rated');
    }

    public function scopeInternal($query)
    {
        return $query->where('call_flow', 'sip_to_sip');
    }

    public function scopeExternal($query)
    {
        return $query->whereIn('call_flow', ['sip_to_trunk', 'trunk_to_sip']);
    }
}

File: python-services/shared/models/call_record.py
from sqlalchemy import Column, BigInteger, String, Enum, Numeric, Integer, DateTime
from shared.database import Base


class CallRecord(Base):
    __tablename__ = "call_records"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    uuid = Column(String(36), nullable=True)
    sip_account_id = Column(BigInteger, nullable=True)
    user_id = Column(BigInteger, nullable=False)
    reseller_id = Column(BigInteger, nullable=True)
    call_flow = Column(
        Enum("sip_to_trunk", "sip_to_sip", "trunk_to_sip", "trunk_to_trunk"),
        nullable=False,
    )
    call_type = Column(
        Enum("regular", "broadcast"),
        default="regular",
    )
    caller = Column(String(40), nullable=False)
    callee = Column(String(40), nullable=False)
    caller_id = Column(String(80), nullable=True)
    src_ip = Column(String(45), nullable=True)
    dst_ip = Column(String(45), nullable=True)
    incoming_trunk_id = Column(BigInteger, nullable=True)
    outgoing_trunk_id = Column(BigInteger, nullable=True)
    did_id = Column(BigInteger, nullable=True)
    broadcast_id = Column(BigInteger, nullable=True)
    destination = Column(String(100), default="")
    matched_prefix = Column(String(20), default="")
    rate_per_minute = Column(Numeric(10, 6), default=0)
    connection_fee = Column(Numeric(10, 6), default=0)
    rate_group_id = Column(BigInteger, nullable=True)
    call_start = Column(DateTime, nullable=False)
    call_end = Column(DateTime, nullable=True)
    duration = Column(Integer, default=0)
    billsec = Column(Integer, default=0)
    billable_duration = Column(Integer, default=0)
    total_cost = Column(Numeric(10, 4), default=0)
    reseller_cost = Column(Numeric(10, 4), default=0)
    trunk_cost = Column(Numeric(10, 4), default=0)
    disposition = Column(
        Enum("ANSWERED", "NO ANSWER", "BUSY", "FAILED", "CANCEL"),
        nullable=True,
    )
    hangup_cause = Column(String(50), nullable=True)
    status = Column(
        Enum("in_progress", "rated", "charged", "failed", "unbillable", "completed"),
        default="in_progress",
    )
    ast_channel = Column(String(80), nullable=True)
    ast_dstchannel = Column(String(80), nullable=True)
    ast_context = Column(String(40), nullable=True)
    rated_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, nullable=True)
